### PR TITLE
Execution time for trades table

### LINKED ISSUE #424: PF transactions list: incorrect Execution time and Execution price values
Discussion https://cowservices.slack.com/archives/C0361CDG8GP/p1680695597411899

1. Open https://explorer-dev-git-partial-fills-table-1-cowswap.vercel.app/orders/0x4a3927813f512fcc6ba710ee44d2a9204d21e997c5c122a60d1d4a2106fdd848042b32ac6b453485e357938bdc38e0340d4b927661c2639b?tab=fills
2. Check Execution time and Execution price values

**AR**: 
Execution price = Sell amount
Execution time = current time

ER: all values are correct (execution time = block time, price is also correct)

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

Closes #424

## Changes
1. For trades added fetching of blocks timestamps using `web3.eth.getBlock()`
2. The loading shimmer is displayed while blocks timestamps are loading
4. The trades table must be displayed even if we couldn't load timestamps (due to separated `useEffect()`)

image

image

### CODE REVIEW

**Comment by elena-zh** [comment]
Hey @shoom3301 , great job!
One nitpick: when the shimmering effect is running, revert price option goes to the default state. Could we please not revert it back at this time? 
Also, the screen is blinking when the data is updated. See the video: https://www.loom.com/share/e21b768d5e2d404d9e50cc0102ae89bb

Could you please enhance this behavior a bit?

Thank you!

File: src/hooks/useOperatorTrades.ts [hunk 1]
```diff
@@ -1,68 +1,33 @@
...
-  const [isLoading, setIsLoading] = useState(false)
-  const [error, setError] = useState<UiError>()
-  const [trades, setTrades] = useState<Trade[]>([])
+async function fetchTradesTimestamps(rawTrades: RawTrade[]): Promise<TradesTimestamps> {
+  const requests = rawTrades.map(({ txHash, blockNumber }) => {
+    return web3.eth.getBlock(blockNumber).then(res => ({
+      txHash,
+      timestamp: +res.timestamp
+    }))
+  })
```
Comment (alfetopito) [inline src/hooks/useOperatorTrades.ts:22]: Can this be cached in memory?
Now it's being updated at every trades list update, which is not really needed. Block info will not change

Comment (shoom3301) [reply src/hooks/useOperatorTrades.ts:22]: It must be, thanks a lot!

**Comment by shoom3301** [comment]
> Also, can you add the full ISO date string to the date on hover? Actually, on click. See how it behaves on the details tab image

@alfetopito It's already integrated, as I can see there is the same components as in details tab

**Comment by shoom3301** [comment]
> Hey @shoom3301 , great job! One nitpick: when the shimmering effect is running, revert price option goes to the default state. Could we please not revert it back at this time? Also, the screen is blinking when the data is updated. See the video: https://www.loom.com/share/e21b768d5e2d404d9e50cc0102ae89bb
> 
> Could you please enhance this behavior a bit?
> 
> Thank you!

@alfetopito Can you reproduce it? I can't

**Comment by alfetopito** [comment]
> > Hey @shoom3301 , great job! One nitpick: when the shimmering effect is running, revert price option goes to the default state. Could we please not revert it back at this time? Also, the screen is blinking when the data is updated. See the video: https://www.loom.com/share/e21b768d5e2d404d9e50cc0102ae89bb
> > Could you please enhance this behavior a bit?
> > Thank you!
> 
> @alfetopito Can you reproduce it? I can't

I could do it before, but now that the refreshing is gone, no more.

> > Also, can you add the full ISO date string to the date on hover? Actually, on click. See how it behaves on the details tab image
> 
> @alfetopito It's already integrated, as I can see there is the same components as in details tab

🤦 I hovered over and saw nothing, forgot to click

**Review by alfetopito — APPROVED** [review]

**Comment by nenadV91** [comment]
I can see this what Elena wrote "Also, the screen is blinking when the data is updated"
https://www.loom.com/share/b65daf10fd6d481e9e855dfc07830bf0

It seems that on each update there is a short moment where the whole page content reloads and it shows initial loader
Screenshot 2023-04-06 at 11 50 11

File: src/hooks/useOperatorTrades.ts [hunk 2]
```diff
@@ -1,68 +1,44 @@
...
+    }
 
-    try {
-      let trades: TradeMetaData[] = []
-      if (orderId) {
-        trades = await getTrades({ networkId, orderId })
-      } else if (owner) {
-        trades = await getTrades({ networkId, owner })
-      }
+    return web3.eth.getBlock(blockNumber).then(res => {
```
Comment (anxolin) [inline src/hooks/useOperatorTrades.ts:26]: I think we have different linter rules. Not sure if mine or yours is the good one 
image

File: src/hooks/useOperatorTrades.ts [hunk 3]
```diff
@@ -1,68 +1,44 @@
...
- * When no filter is given, fetches all trades for current network
- */
-export function useTrades(params: Params): Result {
-  const { owner, orderId } = params
+type TradesTimestamps = { [txHash: string]: number }
 
-  const [isLoading, setIsLoading] = useState(false)
-  const [error, setError] = useState<UiError>()
-  const [trades, setTrades] = useState<Trade[]>([])
+const tradesTimestampsCache: { [blockNumber: number]: number } = {}
```
Comment (anxolin) [inline src/hooks/useOperatorTrades.ts:16]: Are we caching forever?

File: src/hooks/useOperatorTrades.ts [hunk 4]
```diff
@@ -1,68 +1,44 @@
...
 
-  // Here we assume that we are already in the right network
-  // contrary to useOrder hook, where it searches all networks for a given orderId
-  const networkId = useNetworkId()
+async function fetchTradesTimestamps(rawTrades: RawTrade[]): Promise<TradesTimestamps> {
+  const requests = rawTrades.map(({ txHash, blockNumber }) => {
+    const cachedValue = tradesTimestampsCache[blockNumber]
```
Comment (anxolin) [inline src/hooks/useOperatorTrades.ts:20]: nitpick: becasue the cache, caches numbers and not Promise<number>, trades executed at the same block won't hit the cache.

I don't think is a big issue, feel free to ignore.

File: src/hooks/useOperatorTrades.ts (hunk 3 above)
Comment (anxolin) [reply src/hooks/useOperatorTrades.ts:16]: I don't care a lot, since a refresh solves it...

File: src/hooks/useOperatorTrades.ts [hunk 5]
```diff
@@ -72,44 +48,77 @@ export function useOrderTrades(order: Order | null): Result {
...
     },
-    [],
+    [order],
   )
 
+  // Fetch blocks timestamps for trades
+  useEffect(() => {
+    setTradesTimestamps({})
+
+    fetchTradesTimestamps(rawTrades).then(setTradesTimestamps).catch(error => {
```
Comment (anxolin) [inline src/hooks/useOperatorTrades.ts:89]: same issue with linter here 
image

mine likes it like this, which i think is nicer in this case 

image

**Review by anxolin — APPROVED** [review]
I really like the PR, and we should add it. 

However, I also think we should convince backend to add this in the returned data, and delete it later. 

Also this will benefit other apps including CoW Swap. In my opinion is important for the API

File: src/hooks/useOperatorTrades.ts (hunk 3 above)
Comment (shoom3301) [reply src/hooks/useOperatorTrades.ts:16]: Timestamp of a block is constant, so it's ok

**Review by alfetopito — COMMENTED** [review]
The whole page doesn't blink anymore!

Now only the tabs do 😅 

https://user-images.githubusercontent.com/43217/230359333-a5979472-3b10-44cd-a3a8-d374a8a3a6f5.mov

File: src/hooks/useOperatorTrades.ts (hunk 2 above)
Comment (shoom3301) [reply src/hooks/useOperatorTrades.ts]: Thanks! Fixed

File: src/hooks/useOperatorTrades.ts (hunk 4 above)
Comment (shoom3301) [reply src/hooks/useOperatorTrades.ts:20]: Very good point! Actually, I thought about it but didn't implement it for some reason.
Fixed.

File: src/hooks/useOperatorTrades.ts (hunk 5 above)
Comment (shoom3301) [reply src/hooks/useOperatorTrades.ts]: Fixed, and added `.editorconfig`

File: src/hooks/useOperatorTrades.ts (hunk 3 above)
Comment (anxolin) [reply src/hooks/useOperatorTrades.ts]: yes, my point was more like we never garbage collect anything, but as i said above, doesn't matter too much, a refresh will make our app forget about the cache

😈 Making a bit of trouble:
> Timestamp of a block is constant, so it's ok

This is not true necessarily (reorgs). We can ignore this anyways, I'm not concerned with this at all

**Comment by shoom3301** [comment]
@alfetopito thanks! Fixed

**Review by nenadV91 — APPROVED** [review]

**Comment by elena-zh** [comment]
LGTM

**Pull request merged**